src/i2s.rs

```diff
@@ -0,0 +1,1835 @@
+    #[derive(Clone, Copy, Eq, PartialEq)]
+    pub enum SlotBitWidth {
+        /// Slot bit width is automatically set to the data bit width.
+        Auto,
+
+        /// Slot bit width is 8 bits.
+        Bits8,
+
+        /// Slot bit width is 16 bits.
+        Bits16,
+
+        /// Slot bit width is 24 bits.
+        Bits24,
+
+        /// Slot bit width is 32 bits.
+        Bits32,
+    }
+
+    impl Default for SlotBitWidth {
+        #[inline(always)]
+        fn default() -> Self {
+            Self::Auto
+        }
+    }
+
+    impl SlotBitWidth {
+        /// Convert this to the underlying SDK type.
+        #[inline(always)]
+        pub(crate) fn as_sdk(&self) -> i2s_slot_bit_width_t {
+            match self {
+                Self::Auto => 0,
+                Self::Bits8 => 8,
+                Self::Bits16 => 16,
+                Self::Bits24 => 24,
+                Self::Bits32 => 32,
+            }
+        }
+    }
+
+    impl TryFrom<u32> for SlotBitWidth {
+        type Error = EspError;
+
+        fn try_from(value: u32) -> Result<Self, Self::Error> {
+            match value {
+                0 => Ok(Self::Auto),
+                8 => Ok(Self::Bits8),
+                16 => Ok(Self::Bits16),
+                24 => Ok(Self::Bits24),
+                32 => Ok(Self::Bits32),
+                _ => Err(EspError::from(ESP_ERR_INVALID_ARG).unwrap()),
+            }
+        }
+    }
+
+    /// I2S channel slot mode.
+    #[derive(Clone, Copy, Eq, PartialEq)]
+    pub enum SlotMode {
+        /// Mono mode:
+        /// * When transmitting, transmit the same data in all slots.
+        /// * When receiving, only receive data from the first slot.
+        Mono,
+
+        /// Stereo mode:
+        /// * When transmitting, transmit different data in each slot.
+        /// * When receiving, receive data from all slots.
+        Stereo,
+    }
+
+    impl Default for SlotMode {
+        #[inline(always)]
+        fn default() -> Self {
+            Self::Stereo
+        }
+    }
+
+    impl SlotMode {
+        /// Convert this to the underlying SDK type.
+        #[inline(always)]
+        pub(crate) fn as_sdk(&self) -> i2s_slot_mode_t {
+            match self {
+                Self::Mono => 0,
+                Self::Stereo => 1,
+            }
+        }
+    }
+
+    pub trait MaybePin {
+        fn pin(&self) -> i32;
+    }
+
+    impl MaybePin for () {
+        fn pin(&self) -> i32 {
+            -1
+        }
+    }
+
+    impl<T: Pin> MaybePin for PeripheralRef<'static, T> {
+        fn pin(&self) -> i32 {
+            <T as Pin>::pin(self)
+        }
+    }
+}
+
+pub trait I2s: Send {
+    fn port() -> i2s_port_t;
+}
+
+/// The inner details about an I2S driver. This is allocated separately to prevent moving the atomics around.
+struct I2sDriverInternal {
+    i2s: u8,
+    config: config::Config,
+    rx_chan_handle: i2s_chan_handle_t,
+    tx_chan_handle: i2s_chan_handle_t,
+    rx_chan_available: AtomicBool,
```

Comment:
I'm looking at the code and still do not understand what is protected by these atomics?

Reply:
This keeps track whether the rx/tx channel is currently open and will refuse to create a new channel. With I2S, you can open an RX channel, close it, open it again with a different configuration, etc.

Reply:
No longer an issue. Channel structs are gone.